# Assignment the First

## Part 1
1. Be sure to upload your Python script. Provide a link to it here:

| File name | label | Read length | Phred encoding |
|---|---|---|---|
| 1294_S1_L008_R1_001.fastq.gz |read1  | 101 | Phred+33 |
| 1294_S1_L008_R2_001.fastq.gz |index1  | 8 | Phred+33 |
| 1294_S1_L008_R3_001.fastq.gz |index2  | 8 | Phred+33 |
| 1294_S1_L008_R4_001.fastq.gz |read2 | 101 | Phred+33 |

2. Per-base NT distribution
    1. Use markdown to insert your 4 histograms here.
    2. I used this website for my answer justification: https://www.illumina.com/Documents/products/technotes/technote_Q         Scores.pdf#:~:text=Illumina%20sequencing%20Q%20scores%20are%20highly%20accurate.%20This,MiSeq%20vs%20HiSeq%202000E.coli%20K12%20MG1655%20Data%20Comparison
       Q scores can reveal how much of the data from a given run is usable in a resequencing or assembly experiment. Sequencing data with lower quality scores can result in significant portion of the reads being unusable, resulting in wasted time and expense.
    4. **YOUR ANSWER HERE**
    
## Part 2
1. Define the problem
2. Describe output
3. Upload your [4 input FASTQ files](../TEST-input_FASTQ) and your [>=6 expected output FASTQ files](../TEST-output_FASTQ).
4. Pseudocode
5. High level functions. For each function, be sure to include:
    1. Description/doc string
    2. Function headers (name and parameters)
    3. Test examples for individual functions
    4. Return statement
